### ISSUE TITLE
Strange behavior of :

### ISSUE THREAD

**Issue by mariari** [FStarLang/FStar#1905]
```fstar
val test : int
let test =
    let ans : int = 5 in
    ans : 2

val test' : int
let test' =
    let ans : int = 5 in
    ans : int

val test'' : int
let test'' =
    let ans : string = "hello" in
    ans : 3
```

In the above example, test turns out to be 2.

Generating the OCaml output shows that when one write `x : y` in an expression like that, the x is ignored and it simply returns y.

```ocaml
let (test : Prims.int) = let ans = (Prims.of_int (5)) in (Prims.of_int (2))
```
Knowing this, `test'` fails to verify for the expected reason listed below

```fstar
(Error) Expected expression of type "int"; got expression "Prims.int" of type "eqtype"
```

`test''` verifies, showing that `:` in this case does not care about the type on the left side.

This seems like invalid behavior for `:`

**Comment by mtzguido** [comment]
These names are completely ignored by the desugarer if they appear in this fashion.

I tried to rework the parser to forbid this, but couldn't manage. The issue is that the parser rule in question is also used to take the names from binders in an arrow type such as `x:int -> Lemma (x*x >= 0)`. These rules look like `e1=term RARROW term` (instead of say `binder RARROW term`), where the semantic action will inspect `e1` to see if it is a `NamedTyp` and act accordingly. Perhaps someone more familiar with the parser can refactor it to not need these `NamedTyp` nodes.

Another approach is to just fail on desugaring if a `NamedTyp` makes it to a term position. See this commit 8f450aaec612488247b914ffc8d51b22eefe3cbe. It should be roughly the same, except for 1) the slightly different error message and 2) the fact that `--print` will succeed on a file with a bad usage of `:`, since it parses properly. (Though we could make the printer fail on `NamedTyp` too...)

The patch introduces a few (easy) regressions that I'm fixing. I'll say a bit about this on the meeting tomorrow.

**Comment by mtzguido** [comment]
Reading this issue again perhaps the bug wasn't clear: using `a : b` is not ascribing the type `b` to `a`, but instead "naming" `b` with name `a`. These names are completely ignored unless they appear in a binder `x:int -> ...` or an argument `f (x:a) = ...` etc. The solution we discussed was to warn when these are being ignored, like follows:

```
Bug1905.fst(6,4-6,7): (Warning 331) This name is being ignored
Bug1905.fst(11,4-11,7): (Warning 331) This name is being ignored
Bug1905.fst(16,4-16,7): (Warning 331) This name is being ignored
Bug1905.fst(11,10-11,13): (Error 189) Expected expression of type "Prims.int"; got expression "Prims.int" of type "Prims.eqtype"
```

If you wanted to ascribe a type, the syntax is `e < :t`.

Fix incoming to master.

**Comment by mtzguido** [comment]
Fixed by 8f450aaec612488247b914ffc8d51b22eefe3cbe and 230b4e2e6ee4b278a271afff2e03b77d021ec9b3
